--- a/target/linux/ramips/dts/GL-MT300N-V2.dts
+++ b/target/linux/ramips/dts/GL-MT300N-V2.dts
@@ -79,7 +79,7 @@
 
 &wmac {
 	status = "okay";
-	ralink,mtd-eeprom = <&factory 0x04000>;
+	ralink,mtd-eeprom = <&factory 0x8000>;
 };
 
 &i2c {
@@ -133,3 +133,12 @@
 &ohci {
 	status = "okay";
 };
+
+&pcie {
+	status = "okay";
+
+	pcie-bridge {
+		mediatek,mtd-eeprom = <&factory 0>;
+		mediatek,2ghz = <0>;
+	};
+};
